forge/csrc/passes/commute_utils.cpp

```diff
@@ -898,18 +898,213 @@ bool can_commute_past_op(
     return (is_elementwise(op) and op->op_name() != "interleave") or is_quantization_ops(op);
 }
 
+/**
+ * @brief Updates the attributes and named attributes of "select" operation with new values for dimension, begin,
+ * length, and stride.
+ */
+void update_select_attr(
+    graphlib::OpNode *select_op,
+    int select_dim,
+    std::optional<int> begin,
+    std::optional<int> length,
+    std::optional<int> stride)
+{
+    log_trace("op name = {}", select_op->name());
+
+    if (select_op->op_name() != "select")
+    {
+        return;
+    }
+    assert(select_op->op_name() == "select" && "update_select_attr called for a non-select operation");
+
+    graphlib::OpType::Attrs named_attrs = select_op->named_attrs();
+    std::vector<graphlib::OpType::Attr> attr = select_op->op_attrs();
+    log_trace("current_attrs size = {}", attr.size());
+
+    assert(attr.size() == 4 && "Invalid attribute size for select operation: expected 4");
+
+    std::get<int>(attr[0]) = select_dim;
+    named_attrs["select_dim"] = select_dim;
+
+    if (begin.has_value())
+    {
+        std::get<int>(attr[1]) = begin.value();
+        named_attrs["begin"] = begin.value();
+    }
+    else
+    {
+        named_attrs["begin"] = std::get<int>(attr[1]);
+    }
+
+    if (length.has_value())
+    {
+        std::get<int>(attr[2]) = length.value();
+        named_attrs["length"] = length.value();
+    }
+    else
+    {
+        named_attrs["length"] = std::get<int>(attr[2]);
+    }
+
+    if (stride.has_value())
+    {
+        std::get<int>(attr[3]) = stride.value();
+        named_attrs["stride"] = stride.value();
+    }
+    else
+    {
+        named_attrs["stride"] = std::get<int>(attr[3]);
+    }
+
+    select_op->overwrite_op_named_attrs(attr, named_attrs);
+    log_trace(
+        "Updated select operation {}: select_dim = {}, begin = {}, length = {}, stride = {}",
+        select_op->name(),
+        select_dim,
+        begin.value_or(std::get<int>(attr[1])),
+        length.value_or(std::get<int>(attr[2])),
+        stride.value_or(std::get<int>(attr[3])));
+}
+
+/**
+ * @brief Updates the attributes and named attributes of concat operation with new dimension.
+ */
+void update_concat_attr(graphlib::OpNode *concatenate, int dim)
+{
+    if (concatenate->op_name() != "concatenate")
+        return;
```

Comment:
Also, we we need asserts here and in similar places? 

Just clarify to me if this is expected for one reason or another :))

Reply:
After checking for all ops whether this check is needed, it is observed that when this assert is added to update_reshape_attr
`TT_ASSERT(reshape->op_name() == "reshape", "update_reshape_attr called for a non-reshape operation");`
All the transpose related gtests are failing with this assertion
```
Note: Google Test filter = EraseInverseOps.erase_transpose
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from EraseInverseOps
[ RUN      ] EraseInverseOps.erase_transpose
2024-11-27 11:39:31.504 | INFO     | Always          - op name3 = post_transpose_operand_commute_clone7
unknown file: Failure
C++ exception with description "TT_ASSERT @ /proj_sw/user_dev/mramanathan/lab89_nov26_forge/tt-forge-fe/forge/csrc/passes/commute_utils.cpp:1089: reshape->op_name() == "reshape"
info:
update_reshape_attr called for a non-reshape operation
backtrace:
 --- tt::passes::update_reshape_attr(tt::graphlib::OpNode*, tt::graphlib::Shape)
 --- tt::passes::commute_and_bypass(tt::graphlib::Graph*, std::vector<tt::graphlib::Node*, std::allocator<tt::graphlib::Node*> > const&)
 --- tt::passes::erase_inverse_ops(tt::graphlib::Graph*)
 --- ./build/forge/csrc/test/test_passes(+0x4f90a) [0x55738e7d190a]
 --- ./build/forge/csrc/test/test_passes(+0xe3bcf) [0x55738e865bcf]
 --- ./build/forge/csrc/test/test_passes(+0xd8046) [0x55738e85a046]
 --- ./build/forge/csrc/test/test_passes(+0xd81c5) [0x55738e85a1c5]
 --- ./build/forge/csrc/test/test_passes(+0xd8779) [0x55738e85a779]
 --- ./build/forge/csrc/test/test_passes(+0xd8e7f) [0x55738e85ae7f]
 --- ./build/forge/csrc/test/test_passes(+0xe4197) [0x55738e866197]
 --- ./build/forge/csrc/test/test_passes(+0xd828c) [0x55738e85a28c]
 --- ./build/forge/csrc/test/test_passes(+0x2c536) [0x55738e7ae536]
 --- /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7fa075803d90]
 --- /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80) [0x7fa075803e40]
 --- ./build/forge/csrc/test/test_passes(+0x2c425) [0x55738e7ae425]
" thrown in the test body.
[  FAILED  ] EraseInverseOps.erase_transpose (3339 ms)
[----------] 1 test from EraseInverseOps (3339 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (3339 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] EraseInverseOps.erase_transpose

```
as the op named being received is transpose and this gtest was previously passing since in the original implementation , it is being returned if op is not reshape as mentioned [here](https://github.com/tenstorrent/tt-forge-fe/blob/b09f1cb4da1595c34bde588d5239f8b38f65d540/forge/csrc/passes/commute_utils.cpp#L903).

Meanwhile, for all ops just adding TT_ASSERT statements seems to be enough.

Reply:
I see, thanks for the details. Eventually, we'll have to restructure these as well. For not, this is good enough. Thanks for clearing these out Menakshi! :))